File: Python/101.Symmetric_Tree.py
# 101. Symmetric Tree.

# Topic: Tree, Depth-First Search, Breadth-First Search, Binary.

"""
## Task:
---------
Given the root of a binary tree, check whether it is a mirror of itself (i.e., symmetric around its center).

Example 1:
Input: root = [1,2,2,3,4,4,3]
Output: true

Example 2:
Input: root = [1,2,2,null,3,null,3]
Output: false

Constraints:
The number of nodes in the tree is in the range [1, 1000].
-100 <= Node.val <= 100

Follow up: Could you solve it both recursively and iteratively?


## Testcase:
-------------
[1,2,2,3,4,4,3]
[1,2,2,null,3,null,3]


## Code:
------------
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        

"""
# Solution:

## Recursive Implementation:

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        # If root is None, the tree is symmetric
        if not root:
            return True
        
        # Recursive helper function to check if two trees are mirror images
        def isMirror(left, right):
            # Both are None, symmetric
            if not left and not right:
                return True
            # One is None and the other is not, not symmetric
            if not left or not right:
                return False
            # Check values are the same and recursively check subtrees
            return left.val == right.val and isMirror(left.left, right.right) and isMirror(left.right, right.left)
        
        # Start the recursive checking from the root
        return isMirror(root.left, root.right)


## Iterative Implementation:
    
    def isSymmetricIterative(self, root: Optional[TreeNode]) -> bool:
        # Tree is symmetric if root is None
        if not root:
            return True
        
        # Initialize a queue for level-order traversal
        queue = [root.left, root.right]
        while queue:
            # Pop two nodes for comparison
            left = queue.pop(0)
            right = queue.pop(0)
            
            # Continue if both are None
            if not left and not right:
                continue
            # If one is None or values are different, tree is not symmetric
            if not left or not right or left.val != right.val:
                return False
            
            # Enqueue children for the next level of comparison
            queue.append(left.left)
            queue.append(right.right)
            queue.append(left.right)
            queue.append(right.left)
        
        # If all comparisons passed, the tree is symmetric
        return True


# Description:
'''
To solve the problem of determining whether a binary tree is symmetric, we can approach it both recursively and iteratively. 
The key idea is to compare the left and right subtrees of the tree and check if they are mirror images of each other.

### Recursive Approach:

1. **Base Case**: If the root is `None`, the tree is symmetric.
2. **Symmetry Check Function**: Define a helper function (e.g., `isMirror`) that takes two nodes as input and checks 
     if they are mirrors of each other.
3. **Comparisons**: In `isMirror`, check if:
   - Both nodes are `None` (symmetric case).
   - One of them is `None` and the other isn't (asymmetric case).
   - The values of the two nodes are the same, and the left subtree of one is the mirror of the right subtree of the 
     other, and vice versa.
4. **Initial Call**: Call `isMirror` with the left and right children of the root.

### Iterative Approach:

1. **Queue for Level Order Traversal**: Use a queue to perform a level-order traversal of the tree, comparing nodes 
     at each level.
2. **Enqueue**: Initially, enqueue the left and right children of the root.
3. **Level-wise Symmetry Check**: At each step, dequeue two nodes and compare them. Then enqueue the children of 
     these nodes in a specific order: left child of the first node with the right child of the second node, and 
     right child of the first node with the left child of the second node.
4. **Continue Until Queue is Empty**: If at any point the comparison fails, return `false`. If the queue is empty
     and all comparisons have passed, return `true`.

### In the code:
----------------
- `isSymmetric` implements the recursive approach.
- `isSymmetricIterative` implements the iterative approach using a queue. 
- Both methods start by checking if the root is `None`. If it is, the tree is symmetric by definition.
- The recursive `isMirror` function and the loop in `isSymmetricIterative` perform the actual symmetry checks as described.

'''

File: Cpp/101. Symmetric Tree.cpp
// 101. Symmetric Tree.


// Topic: Tree, Depth-First Search, Breadth-First Search, Binary.

/*
## Task:
---------
Given the root of a binary tree, check whether it is a mirror of itself (i.e., symmetric around its center).

Example 1:
Input: root = [1,2,2,3,4,4,3]
Output: true

Example 2:
Input: root = [1,2,2,null,3,null,3]
Output: false

Constraints:
The number of nodes in the tree is in the range [1, 1000].
-100 <= Node.val <= 100

Follow up: Could you solve it both recursively and iteratively?


## Testcase:
-------------
[1,2,2,3,4,4,3]
[1,2,2,null,3,null,3]


## Code:
------------
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 *//*
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        
    }
};     

*/
// Solution 1:

// The recursive solution can be implemented:

class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (!root) return true; // An empty tree is symmetric
        return isMirror(root->left, root->right);
    }

private:
    bool isMirror(TreeNode* left, TreeNode* right) {
        if (!left && !right) return true; // Both nodes are null
        if (!left || !right) return false; // Only one of the nodes is null
        if (left->val != right->val) return false; // Node values are different

        // Check if the outer and inner pairs are symmetric
        return isMirror(left->left, right->right) && isMirror(left->right, right->left);
    }
};


/*
To check whether a binary tree is symmetric around its center, we can compare the left and right subtrees to ensure they are mirror images of each other. This can be done both recursively and iteratively. 

### Recursive Solution:
The recursive approach involves a helper function that takes two nodes to compare. Initially, it will be called with the left and right children of the root. The function will then check the following conditions for symmetry:
1. Both nodes are null, which is a base case for symmetry.
2. One node is null and the other is not, which breaks symmetry.
3. The values of the two nodes are different, which breaks symmetry.
4. Recursively check the outer and inner pairs of the subtrees: the left child of the left subtree with the right child of the right subtree, and the right child of the left subtree with the left child of the right subtree.

*/

// Solution2:
// The iterative solution using a queue:

```cpp
#include <queue>

class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        std::queue<TreeNode*> q;
        q.push(root);
        q.push(root);

        while (!q.empty()) {
            TreeNode* t1 = q.front(); q.pop();
            TreeNode* t2 = q.front(); q.pop();

            if (!t1 && !t2) continue;
            if (!t1 || !t2) return false;
            if (t1->val != t2->val) return false;

            q.push(t1->left);
            q.push(t2->right);
            q.push(t1->right);
            q.push(t2->left);
        }

        return true;
    }
};

/*
### Iterative Solution:

The iterative approach can use a queue or stack to compare the nodes at each level. The idea is to enqueue or push two nodes at 
a time and then compare them, ensuring that they are symmetric. This process is then repeated for the children of these nodes, 
enqueuing them in opposite order (left child of the left node with right child of the right node, and vice versa).

Both solutions effectively check if the binary tree is symmetric around its center by comparing the left and right subtrees for 
mirror symmetry, either recursively or iteratively.

*/


